### PR TITLE
Full additive joint

### PR DESCRIPTION
**Pull request by evvahu**
This PR contains the full additive model as a joint model and a test unit to test it. 
First, the adjective and the noun get transformed by individual weights. Then, the adjective and noun are linearly transformed separately: 1) for attribute prediction 2) for phrase reconstruction. 
These two vectors are summed to form the combined vector. 
Thus, the model has three output vectors.

I also removed redundant embedding normalization in the matrix_joint_model (the tests still works).

### CODE REVIEW

File: ranking_models/full_additive_joint_ranking.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        """
+        :param input_dim: embedding dimension
+        :param dropout_rate: dropout rate for regularization
+        :param normalize_embeddings: whether the composed representation should be normalized to unit length
+        """
+        super(FullAdditiveJointRanker, self).__init__()
+
+        self._general_adj_matrix = nn.Linear(input_dim, input_dim)
+        self._general_noun_matrix = nn.Linear(input_dim, input_dim)
+
```
Comment (Blubberli) [inline ranking_models/full_additive_joint_ranking.py:28]: hmm jetzt haben wir es so:
1) linear transformation with bias of each positional vector (each vector gets transformed with a position-specific matrix)
2) task-specific linear transformation of each vector and combination of the two word vectors via sum
3) combination of task-specific representations via sum

mir ist aufgefallen:
- wir haben verschiedene intialisierungsmethoden (identity matrix für task-specific, random für general). identity matrix Initialisierungen machen vor allen Dingen sinn wenn ich die direkt auf die originalen Vektoren anwende, denn dann kann das Modell, wenn die originalen vectoren schon ohne transformation perfekt wären, nahe an der identity matrix bleiben. also wenn würde vielleicht erst identity und dann random sinn machen, wahrscheinlich aber wäre es gut wenn alle gleich initialisiert, also entweder random oder idenity würden
- da es im originalen identity ist, würde ich jetzt erstmal diese Initialisierung auswählen
- wir haben keine non-linearity (bei den anderen Modellen haben wir eine zwischen den transformationssteps), hier könnte man eine nach der general transformation einbauen
- du hattest jetzt sehr gute Ergebnisse ich habe keine Ahnung ob genau diese Optionen die aktuell da sind, zu diesen guten Ergebnissen führen und zb eine non-linearity nach der general transformation alles kaputt machen würde..also vielleicht kannst du auch einmal auf dem kleinsten Datenset vergleichen wir groß der unterschied mit / ohne relu / mit identity / random init ist...

Comment (Blubberli) [reply ranking_models/full_additive_joint_ranking.py:28]: ansonsten sieht alles perfekt aus, man könnte es also mergen, wenn wir uns hier auf eine Einstellung einigen

Comment (evvahu) [reply ranking_models/full_additive_joint_ranking.py:28]: zu 1.) ja ich wusste nicht ob ich wirklich bias adden soll. Ich hab's einfach dann möglichst analog zum matrix joint model gemacht und da habe ich das eben auch gemacht.
Denkst du, ich soll das ändern? 

meinst du mit _original_ das full_additive model (ohne joint)? Ich weiß nicht was gescheiter ist. bei nn.Linear gibt's, soweit ich herausgefunden habe, keinen Initalisierungs-parameter. Ich könnte also das auch "manuell" machen (matrix multiplication) und falls wir bias behalten noch bias dazu addieren. 

wenn wir das full-additive joint möglichst nahe bei dem von Hartung haben wollen, wieso sollen wir dann eine non-linearity einfügen? Aber klar, ich kann das noch ausprobieren.

Comment (Blubberli) [reply ranking_models/full_additive_joint_ranking.py:28]: dann lassen wir es linear. keine nonlinearity und kein bias. dann also 6 lineare Transformationen, alle mit identity initialisiert

Comment (evvahu) [reply ranking_models/full_additive_joint_ranking.py:28]: ich habe das jetzt gemacht.
Die Frage ist jetzt nur: soll ich das beim Matrix joint model auch noch ändern? und müsste es dann theoretisch beim matrix normalen (nicht joint) auch noch geändert werden?

**Comment by evvahu** [comment]
soo, hoffe stimmt so

**Comment by Blubberli** [comment]
> soo, hoffe stimmt so

kannst du den test noch fixen ? ;)

File: utils/training_utils.py [hunk 2]
```diff
@@ -83,18 +83,20 @@ def init_classifier(config):
...
-        assert config["model"]["task_weights"][0] + config["model"]["task_weights"][
-            1] == 1.0, "the task weights have to sum to 1"
         classifier = MatrixJointRanker(input_dim=config["model"]["input_dim"],
                                        dropout_rate=config["model"]["dropout"],
-                                       normalize_embeddings=config["model"]["normalize_embeddings"])
+                                       normalize_embeddings=config["model"]["normalize_embeddings"],
+                                       non_linearity=config["model"]["non_linearity"])
+    if config["model"]["type"] == "joint_ranking_full_additive":
+        classifier = FullAdditiveJointRanker(input_dim=config["model"]["input_dim"],
+                                             dropout_rate=config["model"]["dropout"],
```
Comment (Blubberli) [inline utils/training_utils.py]: und "dropout rate" ,muss noch weg

**Comment by evvahu** [comment]
> > soo, hoffe stimmt so
> 
> kannst du den test noch fixen ? ;)

nur den Full_additive, oder?
